dali/pipeline/operators/arithmetic/arithmetic.h

```diff
@@ -0,0 +1,268 @@
+  if (expr.GetNodeType() == NodeType::Constant) {
+    expr.SetShape(kernels::TensorListShape<>{{1}});
+    return expr.GetShape();
+  }
+  if (expr.GetNodeType() == NodeType::Tensor) {
+    auto &e = dynamic_cast<ExprTensor&>(expr);
+    expr.SetShape(ws.template InputRef<Backend>(e.GetMappedInput()).shape());
+    return expr.GetShape();
+  }
+  if (expr.GetSubexpressionCount() == 2) {
+    auto left_shape = PropagateShapes<Backend>(expr[0], ws);
+    auto right_shape = PropagateShapes<Backend>(expr[1], ws);
+    expr.SetShape(ShapePromotion(expr.GetOp(), left_shape, right_shape));
+    return expr.GetShape();
+  }
+  DALI_FAIL("Only binary expressions are supported");
+}
+
+/**
+ * @brief Arithmetic operator capable of executing expression tree of elementwise
+ *        arithmetic operations.
+ *
+ * Only expression consisting of one function node with tensor inputs are now supported.
+ *
+ * There are 3 levels for unit of work.
+ * - Thread (CPUBackend) or CUDA kernel invokation (GPUBackend)
+ * - Task - group of tiles to process by thread or CUDA kernel
+ * - Tile - describes a portion of linear buffer, we try to split the amount of work
+ *          evenly into tasks.
+ *
+ * For CPUBackend we have fixed number of threads that get to process a number of tasks,
+ * so the work is evenly distributed. For GPUBackend we pack all tiles into 1 task, to limit
+ * the number of CUDA calls.
+ */
+template <typename Backend>
+class ArithmeticGenericOp : public Operator<Backend> {
+ public:
+  inline explicit ArithmeticGenericOp(const OpSpec &spec) : Operator<Backend>(spec) {
+    expr_ = ParseExpressionString(spec.GetArgument<std::string>("expression_desc"));
+    num_tasks_ = std::is_same<Backend, CPUBackend>::value ? kTaskToThreadRatio * num_threads_ : 1;
+  }
+
+ protected:
+  bool CanInferOutputs() const override {
+    return true;
+  }
+
+  bool SetupImpl(std::vector<OutputDesc> &output_desc, const workspace_t<Backend> &ws) override {
+    output_desc.resize(1);
+
+    result_type_id_ = PropagateTypes<Backend>(*expr_, ws);  // can be done once
+
+    result_shape_ = PropagateShapes<Backend>(*expr_, ws);
+    AllocateIntermediateNodes();
+    exec_order_ = CreateExecutionOrder<Backend>(*expr_, cache_);
+
+    output_desc[0] = {result_shape_, TypeTable::GetTypeInfo(result_type_id_)};
+    std::tie(tile_cover_, tile_range_) = GetTiledCover(result_shape_, kExtent, num_tasks_);
+    return true;
+  }
+
+  using Operator<Backend>::RunImpl;
+  void RunImpl(workspace_t<Backend> &ws) override;
+
+ private:
+  void AllocateIntermediateNodes() {
+    auto &expr = *expr_;
+    bool is_simple_expression = expr.GetSubexpressionCount() == 2 &&
+                                expr[0].GetNodeType() != NodeType::Function &&
+                                expr[1].GetNodeType() != NodeType::Function;
+    DALI_ENFORCE(is_simple_expression, "Complex expression trees are not supported");
```

Comment:
Maybe something like
```suggestion
    DALI_ENFORCE(is_simple_expression, "Complex expression trees are not yet supported. Only expressions containing two function nodes are supported");
```

Reply:
Done